### PR TITLE
rats-tls: add sev attester and verifier

### PR DESCRIPTION
**Pull request by haosanzi**

### CODE REVIEW

File: rats-tls/src/attesters/sev/ttrpc/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            Ok(res) => {
+                let evidence = res.get_evidence();
+                let wrap_command = String::from_utf8(evidence.to_vec())
+                    .and_then(|evidence| Ok(serde_json::from_str::<SevEvidence>(&evidence[..])))
+                    .unwrap();
+
+                match wrap_command {
+                    Ok(wrap_command) => {
+                        debug!("sev evidence is {:#X?}", wrap_command);
+                        return Some(Box::new(wrap_command));
```
Comment (jialez0) [inline rats-tls/src/attesters/sev/ttrpc/src/lib.rs:42]: `return` is no need.

File: rats-tls/src/attesters/sev/ttrpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            0
+        }
+    }
+}
+
+pub fn connect() -> Result<aeb_ttrpc::AebClient> {
+    let addr = format!("{}:{}", SERVER_ADDR, SERVER_PORT);
+    debug!("connect {}", addr);
+
+    let c = Client::connect(&addr[..]).context(format!("failed to connect {}", addr))?;
```
Comment (jialez0) [inline rats-tls/src/attesters/sev/ttrpc/src/lib.rs:93]: Just `&addr` is enough, no need to use `&addr[..]`.

File: rats-tls/cmake/CompilerOptions.cmake [hunk 3]
```diff
@@ -1,9 +1,11 @@
-# Normal and occlum mode
+# Normal, occlum, and sev mode
```
Comment (jiazhang0) [inline rats-tls/cmake/CompilerOptions.cmake:1]: Change to `Normal, occlum and sev modes`.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rats-tls/src/attesters/sev/collect_evidence.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	/* Get guest firmware handle by KVM_HC_VM_HANDLE hypercall */
+	uint32_t guest_handle = do_hypercall(KVM_HC_VM_HANDLE);
+	RTLS_DEBUG("The guest firmware handle is %d\n", guest_handle);
+
+	/* Send retrieve_attestation_evidence request to AEB service through vsock.
+	 * AEB service returns attestation evidence to sev attester.
+	 * The implement of retrieve_attestation_evidence_size, retrieve_attestation_evidence
+	 * is defined in libttrpc.so.
+	 */
+	uint32_t evidence_size = retrieve_attestation_evidence_size(guest_handle);
```
Comment (jiazhang0) [inline rats-tls/src/attesters/sev/collect_evidence.c:46]: Plz add the error check. E.g, what happens if AEB is not started?

File: rats-tls/src/attesters/sev/collect_evidence.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	sev_evidence *s_evidence = retrieve_attestation_evidence(guest_handle, evidence_size);
+	if (!s_evidence) {
+		RTLS_ERR("retrieve attestation_evidence failed\n");
+		return -ENCLAVE_ATTESTER_ERR_INVALID;
+	}
+
+	RTLS_DEBUG("Succeed to generate the sev evidence!\n");
```
Comment (jiazhang0) [inline rats-tls/src/attesters/sev/collect_evidence.c:53]: s/generate/retrieve/

File: rats-tls/src/attesters/sev/collect_evidence.c [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
```
Comment (jiazhang0) [inline rats-tls/src/attesters/sev/collect_evidence.c:54]: Add a check:
```C
if (evidence_size != sizeof(sev_evidence)) {
  // error handling
}
```

File: rats-tls/src/attesters/sev/collect_evidence.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	sev_attestation_evidence_t *sev_report = &evidence->sev;
+	memcpy(sev_report->report, s_evidence, sizeof(*s_evidence));
+	sev_report->report_len = sizeof(sev_evidence);
```
Comment (jiazhang0) [inline rats-tls/src/attesters/sev/collect_evidence.c:57]: change to `sev_report->report_len = evidence_size;`.

File: rats-tls/src/attesters/sev/collect_evidence.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#include <string.h>
+#include "../../verifiers/sev/sevapi.h"
+
+#define KVM_HC_VM_HANDLE 13
+
+/* The API of retrieve_attestation_evidence_size and retrieve_attestation_evidence
+ * is defined in libttrpc.so.
+ */
+extern uint32_t retrieve_attestation_evidence_size(uint32_t guest_handle);
+extern sev_evidence *retrieve_attestation_evidence(uint32_t guest_handle, uint32_t evidence_size);
```
Comment (jiazhang0) [inline rats-tls/src/attesters/sev/collect_evidence.c:19]: rename the type name to sev_evidence_t.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rats-tls/src/core/cpu.c [hunk 9]
```diff
@@ -128,3 +130,36 @@ bool is_tdguest_supported(void)
...
+	if (fd < 0) {
+		close(fd);
+		RTLS_ERR("rdmsr: open error\n");
+		return data;
+	}
+
+	uint64_t size = pread(fd, &data, sizeof data, reg);
+	if (size != sizeof(data)) {
+		close(fd);
+		RTLS_ERR("rdmsr: pread error\n");
```
Comment (jiazhang0) [inline rats-tls/src/core/cpu.c:152]: Change to `RTLS_ERR("failed to read msr %#x\n", reg);`

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rats-tls/src/attesters/sev/ttrpc/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    Ok(aeb_ttrpc::AebClient::new(c))
+}
+
+fn default_ctx() -> ttrpc::context::Context {
+    let ctx = ttrpc::context::with_timeout(0);
+
+    ctx
+}
```
Comment (liangzhou121) [inline rats-tls/src/attesters/sev/ttrpc/src/lib.rs:102]: I think the value can be returned directly, like
fn default_ctx() -> ttrpc::context::Context {
    ttrpc::context::with_timeout(0)
}

File: rats-tls/src/attesters/sev/ttrpc/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+#[no_mangle]
+pub extern "C" fn retrieve_attestation_evidence_size(guest_handle: u32) -> u32 {
+    let req = aeb::RetrieveAttestationEvidenceSizeRequest {
+        guest_handle: guest_handle,
+        unknown_fields: Default::default(),
+        cached_size: Default::default(),
+    };
+
+    let agent_client = connect().context("connect");
```
Comment (liangzhou121) [inline rats-tls/src/attesters/sev/ttrpc/src/lib.rs:70]: Maybe you can change the following snippet codes like this:
let res = connect().context("connect")
    .map_err(|e| {
	    error!("failed to retrieve evidence size from aeb service {:?}", e);
	    0
	})
	.and_then(|agent_client| {
	    let res = agent_client.retrieve_attestation_evidence_size(default_ctx(), &req)
		    .map_err(|e| {
			    error!("failed to retrieve evidence size from aeb service {:?}", e);
			    0
			})
			.and_then(|res| {res.evidence_size})
		OK(res)
	})
	.unwrap();

File: rats-tls/README.md [hunk 12]
```diff
@@ -56,18 +56,25 @@ cmake -DRATS_TLS_BUILD_MODE="tdx" -DBUILD_SAMPLES=on -H. -Bbuild
 make -C build install
 ```
 
+If you want to run SEV instances, please type the following command.
```
Comment (jiazhang0) [inline rats-tls/README.md:59]: s/command/commands/

File: rats-tls/README.md [hunk 13]
```diff
@@ -78,6 +85,13 @@ cd /usr/share/rats-tls/samples
 ./rats-tls-server
 ```
 
+**Notice: special prerequisites for SEV remote attestation in software capability.**
+
+- software capability:
+  - [Kernel supports to get guest firmware handle in guest](../hack/sev-attestation/Linux-5.14-kvm-add-hypercall-to-get-sev-guest-firmware-handle.patch).
+  - [The ccp driver supports to get attestation report with `ioctl` command](../hack/sev-attestation/Linux-5.14-drivers-crypto-ccp-add-SEV_GET_REPORT-command.patch).
+  - Start the [attestation evidence broker](../attestation-evidence-broker/README.md) service in host.
```
Comment (jiazhang0) [inline rats-tls/README.md:93]: sev attester may run in client side, so it is not appropriate to document the steps in "Run RATS TLS server".

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Review by YangLiang3 — APPROVED** [review]

**Review by zhiminghufighting — APPROVED** [review]

File: rats-tls/README.md (hunk 12 above)
Comment (jiazhang0) [reply rats-tls/README.md:59]: Please fix this issue.

File: rats-tls/README.md [hunk 14]
```diff
@@ -91,6 +98,13 @@ Check msr 0x503, return value must be 0:
 sudo rdmsr 0x503s
 ```
 
+**Notice: special prerequisites for SEV remote attestation in software capability.**
+
+- software capability:
+  - [Kernel supports to get guest firmware handle in guest](../hack/sev-attestation/KVM-X86-Introduce-KVM_HC_VM_HANDLE-hypercall.patch).
+  - [The ccp driver supports to get attestation report with `ioctl` command](../hack/sev-attestation/crypto-ccp-Implement-SEV_GET_REPORT-ioctl-command.patch).
+  - Start the [attestation evidence broker](../attestation-evidence-broker/README.md) service in host.
```
Comment (jiazhang0) [inline rats-tls/README.md:106]: Once you move the patches to your AEB repo, then update these info.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rats-tls/README.md [hunk 15]
```diff
@@ -56,18 +56,25 @@ cmake -DRATS_TLS_BUILD_MODE="tdx" -DBUILD_SAMPLES=on -H. -Bbuild
 make -C build install
 ```
 
+If you want to run SEV instances, please type the following commands.
```
Comment (jiazhang0) [inline rats-tls/README.md:59]: change to `SEV(-ES)`.

File: rats-tls/README.md [hunk 16]
```diff
@@ -91,6 +98,15 @@ Check msr 0x503, return value must be 0:
 sudo rdmsr 0x503s
 ```
 
+**Notice: special prerequisites for SEV remote attestation in software capability.**
+
+- software capability:
```
Comment (jiazhang0) [inline rats-tls/README.md:103]: remove this line.

File: rats-tls/README.md [hunk 17]
```diff
@@ -91,6 +98,15 @@ Check msr 0x503, return value must be 0:
...
+  - [Kernel supports to get guest firmware handle in guest](https://github.com/haosanzi/attestation-evidence-broker/blob/master/hack/0001-KVM-X86-Introduce-KVM_HC_VM_HANDLE-hypercall.patch).
```
Comment (jiazhang0) [inline rats-tls/README.md:104]: s/supports/support/

File: rats-tls/README.md [hunk 18]
```diff
@@ -91,6 +98,15 @@ Check msr 0x503, return value must be 0:
...
+  - The ccp driver supports to get attestation report with `ioctl` command, please refer to the following patches for more information.
+    - [Move the implementation of sev_get_attestation_report to ccp driver](https://github.com/haosanzi/attestation-evidence-broker/blob/master/hack/0002-KVM-SVM-move-the-implementation-of-sev_get_attestati.patch).
+    - [Implement SEV_GET_REPORT ioctl command in ccp driver](https://github.com/haosanzi/attestation-evidence-broker/blob/master/hack/0003-crypto-ccp-Implement-SEV_GET_REPORT-ioctl-command.patch)
```
Comment (jiazhang0) [inline rats-tls/README.md:107]: I suggest you create a README under the hack directory to talk about above words, and then refer to it here.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Comment by haosanzi** [comment]
I Will resubmitted to https://github.com/inclavare-containers/rats-tls.

**Comment by haosanzi** [comment]
Resubmitted to https://github.com/inclavare-containers/rats-tls/pull/61.

**Pull request closed without merging**